docs/spec.md

```diff
@@ -140,240 +43,355 @@ stablehlo.func @main(
+statically, because in StableHLO functions are not first-class values) and
+input attributes (also provided statically). The kind of inputs and outputs
+consumed and produced by an op depends on its mnemonic. For example, the `add`
+op consumes 2 input values and produces 1 output value. In comparison, the
+`select_and_scatter` op consumes 3 input values, 2 input functions and
+3 input attributes.
+
+```ebnf
+OpInputFunc ::= '{' Unused FuncInputs ':' FuncBody '}'
+Unused      ::= '^' digit {digit}
+              | '^' letter {letter | digit}
+```
 
-There are five collective ops in StableHLO: `all_gather`, `all_reduce`,
-`all_to_all`, `collective_permute` and `reduce_scatter`. All these ops split
-the processes in the StableHLO grid into **StableHLO process groups** and
-execute a joint computation within each process group, independently from other
-process groups.
+**Input functions** (which are also called **anonymous functions**) are very
+similar to named functions except that: 1) they don't have an identifier (hence
+the name "anonymous"), 2) they don't declare output types (output types are
+inferred from the `return` op within the function).
+
+The syntax for input functions includes a currently unused part (see the
+`Unused` production above) which is there for compatibility with MLIR. In MLIR,
+there is a more general concept of "regions" which can have multiple "blocks"
+of ops connected together via jump ops. These blocks have ids which correspond
+to the `Unused` production, so that they can be distinguished from each other.
+StableHLO doesn't have jump ops, so the corresponding part of MLIR syntax is
+unused (but is still there).
+
+```ebnf
+OpInputAttr      ::= OpInputAttrName '=' OpInputAttrValue
+OpInputAttrName  ::= letter {letter | digit}
+OpInputAttrValue ::= Constant
+```
 
-Within each process group, collective ops may introduce a synchronization
-barrier. Further formalization, e.g. elaborating on when exactly this
-synchronization happens, how exactly the processes arrive at this barrier,
-and what happens if they don't, is TBD.
+**Input attributes** have a name and a value which is one of the supported
+constants. They are the primary way to specify static metadata for program
+elements. For example, the `concatenate` op uses the attribute `dimension` to
+specify the dimension along which its input values are concatenated. Similarly,
+the `slice` op uses multiple attributes like `start_indices` and `limit_indices`
+to specify the bounds that are used to slice the input value.
 
-If the process group involves cross-partition communication, i.e. there are
-processes in the process group whose partition ids are different, then execution
-of the collective op needs a channel, and the collective op must provide a
-positive `channel_id` of type `si64`. Cross-replica communication doesn't need
-channels.
+```ebnf
+OpSignature ::= '(' [ValueType {',' ValueType}] ')' '->' '(' [ValueType {',' ValueType}] ')'
```

Comment:
The following line  (a few lines below) violates this definition:

```
%result = "stablehlo.select_and_scatter"(%operand, %source, %init_value) ({
  ^bb0(%arg0: tensor<i32>, %arg1: tensor<i32>):
    %0 = "stablehlo.compare"(%arg0, %arg1) {
      comparison_direction = #stablehlo<comparison_direction GE>
    } : (tensor<i32>, tensor<i32>) -> tensor<i1>
    "stablehlo.return"(%0) : (tensor<i1>) -> ()
}
```

Specifically `(tensor<i32>, tensor<i32>) -> tensor<i1>` would need to be `(tensor<i32>, tensor<i32>) -> (tensor<i1>)` under the current grammar. Could either resolve by updating the production or the example.